client/src/app/shared/components/os-sort-filter-bar/os-sort-filter-bar.component.ts

```diff
@@ -0,0 +1,212 @@
+     */
+    @Input()
+    public filters: OsFilter[];
+
+    /**
+     * The original, unfiltered datasource (TODO observability)
+     */
+    @Input()
+    public modelRepo: Observable<V[]>;
+
+    @ViewChild('filterMenu')
+    public filterMenu: FilterMenuComponent;
+
+    @ViewChild('sortBottomSheet')
+    public sortBottomSheet: OsSortBottomSheetComponent<V>;
+
+    @Output()
+    public filteredData = new EventEmitter<V[]>();
+
+    private _filteredDataSource = new MatTableDataSource<V>();
+    private _originalData: V[];
+
+    public constructor(
+        private bottomSheet: MatBottomSheet,
+        public translate: TranslateService) {
+
+            this.filterMenu = new FilterMenuComponent(translate);
+            this.filterMenu.activeFilters.subscribe(newFilters => {
+                this.filters = newFilters;
+                this.filterData();
+            });
+    }
+
+    public ngOnInit(): void {
+        if (this.modelRepo){
+            this.modelRepo.subscribe(originalData => {
+                this._originalData = originalData;
+                this.filterData();
+            });
+        }
+    }
+
+    /**
+     * Filter the data, and send to output
+     */
+    private filterData(): void {
+        const newData: V[] = [];
+        for (const item of this._originalData) {
+            let excluded = false;
+            // all data must be part of all active filters
+            for (const filter of this.filters) {
+                if (filter.isActive){
+                    if (!filter.condition) {// TODO avoid active, but empty filter
+                        excluded = true;
+                    } else {
+                        if (filter.condition.indexOf(item[filter.property].toString()) < 0) {
+                            excluded = true;
+                        }
+                    }
+                }
+            }
+            if (!excluded) {
+                newData.push(item);
+            }
+        }
+        if (this.sorting) {
+            newData.sort(this.getSortFn());
+        }
+        this._filteredDataSource.data = newData;
+        this.filteredData.emit(this._filteredDataSource.data);
+    }
+
+    public openSortDropDown(): void {
+        const bottomSheetRef = this.bottomSheet.open(OsSortBottomSheetComponent,
+            { data: this.sorting }
+        );
+        bottomSheetRef.afterDismissed().subscribe(result => {
+            if (result) {
+                // result can be undefined if dismissed without selecting anything
+                this.sorting = result;
+                this.filterData();
+            }
+        });
+    }
+
+    public applySearch(input: string): void { // TODO no effect for now
+        this._filteredDataSource.filter = input;
+        this.filterData();
+    }
+
+    public newFilters(event: OsFilter[]): void {
+        this.filters = event;
+        this.filterData();
+    }
+
+    private getSortFn(): (a: V, b: V) => number {
```

Comment:
-  Sorting can be direct comparison (string different from numbers),  or a localeCompare
- the compared item are a variety of string, the result of a function residing as method of the given item

I currently have a scope problem here: I want to generate a sort function every time the sorting changes , to be used in sort (line 141). If I put the sort function directly there, it is a massive performance hit, because it is generated each sort iteration.

I probably just don't see an easy two-line solution, again.

Reply:
okay, scope is fixed, but performance is still an issue

It is reduced, but unacceptable still:
- 100 generated test motions lead to about 800ms of initial calculating; something's still very off there.